File: 3318-find-x-sum-of-all-k-long-subarrays-i.py
from heapq import heapify, heappop


class Solution:
    def findXSum(self, nums: list[int], k: int, x: int) -> list[int]:
        '''
        We have a loop that goes from 0 to (n - k)
        We have a variable `count` that is a map of the count of each integer
        We also have a variable `result` that stores the result
        In each loop,
            If i == 0,
                We calculate the count of all elements from i to (i + k - 1)
            else,
                we get the previous i, reduce the count for that by 1, and get i + k - 1 and increment the count for it
            Then we add all the elements into a max heap and have a for loop to retrieve the top x elements and add up the sum
            Then add it into the result array
        '''
        result = []
        count = {}
        n = len(nums)
        for i in range(n - k + 1):
            max_heap = []
            sum = 0

            if i == 0:
                for j in range(k):
                    count[nums[j]] = count.get(nums[j], 0) + 1

            else:
                prev = nums[i-1]
                next = nums[i + k - 1]
                count[prev] -= 1
                count[next] = count.get(next, 0) + 1

            for key, val in count.items():
                max_heap.append((-val, -key))

            heapify(max_heap)
            for _ in range(x):
                if max_heap:
                    neg_val, neg_key = heappop(max_heap)
                    sum += -neg_val * -neg_key
            result.append(sum)

        return result


sol = Solution()
nums = [1, 1, 2, 2, 3, 4, 2, 3]
k = 6
x = 2
print(f'output: {sol.findXSum(nums, k, x)}, expected: [6, 10, 12]')

nums = [3, 8, 7, 8, 7, 5]
k = 2
x = 2
print(f'output: {sol.findXSum(nums, k, x)}, expected: [11, 15, 15, 15, 12]')

nums = [9, 2, 2]
k = 3
x = 3
print(f'output: {sol.findXSum(nums, k, x)}, expected: [13]')

File: 3318-find-x-sum-of-all-k-long-subarrays-i.java
import java.util.*;
import java.util.Map.Entry;

class FindXSumOfAllKLongSubarraysI {

    public int[] findXSum(int[] nums, int k, int x) {

        List<Integer> result = new ArrayList<>();
        Map<Integer, Integer> count = new HashMap<>();
        for (int i = 0; i < nums.length - k + 1; i++) {
            Queue<int[]> maxHeap = new PriorityQueue<>((e1, e2) -> {
                if (e1[0] == e2[0]) {
                    return Integer.compare(e2[1], e1[1]);
                }
                return Integer.compare(e2[0], e1[0]);
            });
            int sum = 0;
            if (i == 0) {
                for (int j = 0; j < k; j++) {
                    count.put(nums[j], count.getOrDefault(nums[j], 0) + 1);
                }
            } else {
                int prev = nums[i - 1];
                int next = nums[i + k - 1];
                count.put(prev, count.get(prev) - 1);
                count.put(next, count.getOrDefault(next, 0) + 1);
            }

            for (Entry<Integer, Integer> entry : count.entrySet()) {
                int key = entry.getKey(), val = entry.getValue();
                maxHeap.add(new int[] { val, key });
            }

            for (int j = 0; j < x; j++) {
                if (!maxHeap.isEmpty()) {
                    int[] item = maxHeap.poll();
                    int val = item[0], key = item[1];
                    sum += val * key;
                }
            }
            result.add(sum);
        }
        return result.stream().mapToInt(i -> i).toArray();
    }

    public static void main(String[] args) {
        FindXSumOfAllKLongSubarraysI sol = new FindXSumOfAllKLongSubarraysI();

        System.out.println(
                "Output: " + Arrays.toString(sol.findXSum(new int[] { 1, 1, 2, 2, 3, 4, 2, 3 }, 6, 2)) + ", expected: [6, 10, 12]");
        System.out.println(
                "Output: " + Arrays.toString(sol.findXSum(new int[] { 3, 8, 7, 8, 7, 5 }, 2, 2)) + ", expected: [11, 15, 15, 15, 12]");
        System.out.println("Output: " + Arrays.toString(sol.findXSum(new int[] { 9, 2, 2 }, 3, 3)) + ", expected: [13]");
    }
}
